First week python work

In [120]:

import numpy as np
a=[40,35,40,0,0,40,50,10]
av = np.average(a)

print("Avarege of the array: ",av)

x=np.empty(1000) #after changing 500 to 1000 the results are more stable
for i in range(0,1000): # this loop is building the Bootstrap
    y=np.random.choice(a, size=np.size(a))
    x[i]=np.average(y)

stdx=np.std(x) # standard deviation of the averages of the bootstrap
print("standard deviation for bootstarp: ",stdx)

# for confidence level 95% we can use the following formula
# av+-2*stdx
#print("Confidence level 95%: ")
print("Confidence level 95%: ", av -2*stdx, av+2*stdx)
# for confidence level 99% we can use the following formula
# av+-2.576*stdx
#print("Confidence level 99%: ")
print("Confidence level 99%: ", av -2.576*stdx, av+2.576*stdx)
# for confidence level 90% we can use the following formula
# av+-1.645*stdx
#print("Confidence level 90%: ") 
print("Confidence level 90%: ", av -1.645*stdx, av+1.645*stdx)


print(av -2*stdx, av+2*stdx)

Avarege of the array:  26.875
standard deviation for bootstarp:  6.7315799304008115
Confidence level 95%:  13.411840139198377 40.338159860801625
Confidence level 99%:  9.534450099287508 44.21554990071249
Confidence level 90%:  15.801551014490665 37.94844898550934
13.411840139198377 40.338159860801625


In [121]:
# The number we use to multiply with stdx to get different confidence levels is called the z-score.
# The z-score corresponds to the number of standard deviations a data point is from the mean.
# For a given confidence level, the z-score can be calculated using the inverse cumulative distribution function (CDF) of the standard normal distribution.
# For example:
# - For a 95% confidence level, the z-score is approximately 2. This is because 95% of the data lies within 2 standard deviations from the mean in a standard normal distribution.
# - For a 99% confidence level, the z-score is approximately 2.576.
# - For a 90% confidence level, the z-score is approximately 1.645.
# The formula to calculate the z-score for a given confidence level is:
# z = norm.ppf((1 + confidence_level) / 2)
# where norm.ppf is the percent-point function (inverse of CDF) from the scipy.stats module.

import scipy.stats as stats
from scipy.stats import norm
# Calculate z-scores for different confidence levels
z_95 = norm.ppf((1 + 0.95) / 2)  # 95% confidence level
z_99 = norm.ppf((1 + 0.99) / 2)  # 99% confidence level
z_90 = norm.ppf((1 + 0.90) / 2)  # 90% confidence level

print("Calculated z-scores:")
print("95% confidence level:", z_95)
print("99% confidence level:", z_99)
print("90% confidence level:", z_90)

# Use the calculated z-scores to compute confidence intervals
print("Confidence level 95%: ", av - z_95 * stdx, av + z_95 * stdx)
print("Confidence level 99%: ", av - z_99 * stdx, av + z_99 * stdx)
print("Confidence level 90%: ", av - z_90 * stdx, av + z_90 * stdx)


Calculated z-scores:
95% confidence level: 1.959963984540054
99% confidence level: 2.5758293035489004
90% confidence level: 1.6448536269514722
Confidence level 95%:  13.681345777361766 40.06865422263823
Confidence level 99%:  9.535599156091923 44.21440084390808
Confidence level 90%:  15.802536336366487 37.94746366363351


In [122]:
# calculating the CI (confidence interval) using the percentile method
# x is the array of averages
# 0.5% and 99.5% percentiles are used to calculate the CI in 99% confidence level
# if we want to calculate the CI in 95% confidence level we can use 2.5% and 97.5% percentiles
# if we want to calculate the CI in 90% confidence level we can use 5% and 95% percentiles
# if we want to calculate the CI in 98% confidence level we can use 1% and 99% percentiles

print(np.percentile(x,0.5),np.percentile(x,99.5))

8.75 42.5


חשבו את רווח הסמך לטיפים של שימי ברמת בטחון של 68% בשתי השיטות שלמדנו. 
##video question

In [123]:
# now we want to calculate the CI in 68% confidence level 
# using bothe methods percentile and z-score
# for 68% confidence level we can use 16% and 84% percentiles
print("Confidence level 68%: ", np.percentile(x,16),np.percentile(x,84))
# for 68% confidence level we can use 1*stdx
print("Confidence level 68%: ", av - stdx, av + stdx)



Confidence level 68%:  20.0 33.75
Confidence level 68%:  20.143420069599188 33.60657993040081


1)
bs.std=6.789

X=26.875

CI =  𝑥 ± 1 * bs.std (in case of a 68% CI)

CI=26.875-6.789,26.875+6.789

CI=20.086,33.66


using different bootstarp results to calculate the percentile

using the percentile method

# for 68% confidence level we can use 16% and 84% percentiles
print("Confidence level 68%: ", np.percentile(x,16),np.percentile(x,84))
# for 68% confidence level we can use 1*stdx
print("Confidence level 68%: ", av - stdx, av + stdx)

Confidence level 68%: 20.0 33.75 

Confidence level 68%: 20.143420069599188 33.60657993040081





נסו להעזר ב-AI כדי לפתור את שאלת רווח הסמך. באיזה כלי השתמשתם? 

האם הכלי הצליח ואם לא, מדוע?


using deepseek


Method 1: Standard Formula (Mean ± 1×bs.std)

This assumes a normal distribution of bootstrap samples.
Calculation:
CI=Xˉ±1×bs.std
CI=Xˉ±1×bs.std
Lower bound=26.875−6.789=20.086
Lower bound=26.875−6.789=20.086
Upper bound=26.875+6.789=33.664
Upper bound=26.875+6.789=33.664
Result:
68% CI=[20.086, 33.664]
68% CI=[20.086, 33.664]​
Method 2: Percentile Method (16th & 84th Percentiles)

This uses the empirical distribution of bootstrap samples (no normality assumption).
Key Idea:

    A 68% CI in the percentile method corresponds to the 16th and 84th percentiles of the bootstrap distribution (leaving 16% in each tail).

Given Code Equivalent:
python

print("Confidence level 68%: ", np.percentile(x, 16), np.percentile(x, 84))

Hand Calculation Steps:

    Sort all bootstrap sample means in ascending order.

    Find the 16th percentile:

        Position = 0.16×N0.16×N (where NN = number of bootstrap samples).

        If position is not an integer, interpolate between adjacent values.

    Find the 84th percentile:

        Position = 0.84×N0.84×N.

Example (Hypothetical Data):

Assume sorted bootstrap means (x):
[18.2, 19.5, 20.1, ..., 33.0, 34.5, 35.8]

    For N=1000N=1000:

        16th percentile = Value at position 0.16×1000=1600.16×1000=160 → 20.1

        84th percentile = Value at position 0.84×1000=8400.84×1000=840 → 33.0

Result:
68% CI (Percentile)=[20.1, 33.0]
68% CI (Percentile)=[20.1, 33.0]​

(Note: Exact values depend on your bootstrap sample distribution.)




----------------------------------------------------------------------------------------------------


the ai did the first part correct as it had the values of the video and  bs.std and X  so it was as simple as plugging them into the formula


as for the second part in order to use the percentile method it would need the array of all the bootstrap itteration mean but that datawas not availble so it opted to create hypethetical data that resulted in a different answer from what i got when running the bootstarp

